src/mds/MDSAuthCaps.cc

```diff
@@ -338,8 +339,65 @@ void MDSAuthCaps::set_allow_all()
 				 {}));
 }
 
-bool MDSAuthCaps::parse(std::string_view str, ostream *err)
+/* What should string to be searched next for be -- ", allow" or ",allow"?
+ * Find which of the two is next and return it.
+ *
+ * String to be searched for in MDS caps can't be just "," because when MDS
+ * caps contains multiple GIDs, they too are separated by commas. */
+string_view MDSAuthCaps::_get_search_str(string_view& caps, unsigned int beg)
+{
+  if (caps.find(", allow", beg) != string::npos) {
+    return ", allow";
+  } else if (caps.find(",allow", beg) != string::npos) {
+    return ",allow";
+  } else {
+    return "";
+  }
+}
+
+/* MDS caps don't allow permission "allow w" regardless of what other Ceph
+ * daemons do. In fact it doesn't allow any caps the permission in which isn't
+ * "*" or "all" or doesn't start with "r" or "rw". Check for such permissions,
+ * reject such caps and print a user-friendly error message. */
+bool MDSAuthCaps::_is_perm_correct(string_view& caps, ostream* err)
```

Comment:
Don't we have enough information in L#425 (when parsing fails) about why parsing the MDS auth caps failed? Basically, parse() errors out since (say `w`) is not a valid MDS auth cap, so, isn't that piece of information available for us to deduce the failure rather than handcrafting this routine?

Reply:
Greg wants a special message to printed for this specific case. He said it on the BZ as well as in 1:1. What he said is that `allow w` are allowed on MON caps (IIRC) but not in MDS caps, so CephFS users can make this mistake often. Printing a special message in such a case makes it easy for user decipher their exact mistake.

`fs authorize` already prints this message, if rest of similar interfaces can also print this message, it would be also be uniform.

Reply:
Spoke to @rishabh-d-dave and the plan is to check if the appropriate error can be inferred when parsing fails at L#425 and display that error message accordingly. That would save the handcrafted checks.

Reply:
Got this done. The PR is much much shorter and simpler now. PTAL.